src/lib/mdns/TxtFields.cpp

```diff
@@ -0,0 +1,281 @@
+    }
+    uint8_t ret = 0;
+    for (size_t i = 0; i < ascii.size(); ++i)
+    {
+        ret    = ret << 4;
+        char c = static_cast<char>(ascii.data()[i]);
+        if (c >= '0' && c <= '9')
+        {
+            ret += ascii.data()[i] - static_cast<uint8_t>('0');
+        }
+        // Only uppercase is supported according to spec.
+        else if (c >= 'A' && c <= 'F')
+        {
+            ret += ascii.data()[i] - static_cast<uint8_t>('A') + 0xA;
+        }
+        else
+        {
+            return CHIP_ERROR_INVALID_ARGUMENT;
+        }
+    }
+    *val = ret;
+    return CHIP_NO_ERROR;
+}
+
+size_t GetPlusSignIdx(const ByteSpan & value)
+{
+    // Fist value is the vendor id, second (after the +) is the product.
+    for (int i = 0; i < static_cast<int>(value.size()); ++i)
+    {
+        if (static_cast<char>(value.data()[i]) == '+')
+        {
+            return i;
+        }
+    }
+    return value.size();
+}
+
+} // namespace
+
+uint16_t GetProduct(const ByteSpan & value)
+{
+    size_t plussign = GetPlusSignIdx(value);
+    if (plussign < value.size() - 1)
+    {
+        return MakeU16FromAsciiDecimal(ByteSpan(value.data() + plussign + 1, value.size() - plussign - 1));
+    }
+    return 0;
+}
+
+uint16_t GetVendor(const ByteSpan & value)
+{
+    size_t plussign = GetPlusSignIdx(value);
+    return MakeU16FromAsciiDecimal(ByteSpan(value.data(), plussign));
+}
+
+uint16_t GetLongDisriminator(const ByteSpan & value)
+{
+    return MakeU16FromAsciiDecimal(value);
+}
+
+uint8_t GetAdditionalPairing(const ByteSpan & value)
+{
+    return MakeU8FromAsciiDecimal(value);
+}
+
+uint8_t GetCommissioningMode(const ByteSpan & value)
+{
+    return MakeU8FromAsciiDecimal(value);
+}
+
+// TODO: possibly 32-bit? see spec issue #3226
+uint16_t GetDeviceType(const ByteSpan & value)
+{
+    return MakeU16FromAsciiDecimal(value);
+}
+
+void GetDeviceName(const ByteSpan & value, char * name)
+{
+    size_t len = std::min(value.size(), kMaxDeviceNameLen);
+    memcpy(name, value.data(), len);
+    name[len] = '\0';
+}
+
+void GetRotatingDeviceId(const ByteSpan & value, uint8_t * rotatingId, size_t * len)
+{
+    memset(rotatingId, 0, kMaxRotatingIdLen);
+    *len = 0;
+    // Octet string where each octet is 2 ascii digits representing the hex value
+    // Each is represented by two ascii chars, so must be even number
+    if ((value.size() & 0x1) != 0 || value.size() > (kMaxRotatingIdLen * 2))
+    {
+        return;
+    }
+    for (size_t i = 0; i < value.size(); i += 2)
+    {
+        if (MakeU8FromAsciiHex(ByteSpan(value.data() + i, 2), &rotatingId[i / 2]) != CHIP_NO_ERROR)
+        {
+            *len = 0;
+            memset(rotatingId, 0, kMaxRotatingIdLen);
+            return;
+        }
+        *len = *len + 1;
+    }
+}
+
+uint16_t GetPairingHint(const ByteSpan & value)
+{
+    return MakeU16FromAsciiDecimal(value);
+}
+
+void GetPairingInstruction(const ByteSpan & value, char * pairingInstruction)
```

Comment:
Same comment here about using a common utility, rather than doing more special-casing with baked-in bounds.